File: sach/de3397.cpp
//DE3397.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
const int N = 1e2 + 1;
int a[N][N], vs[N], x[N], y[N], n;
bool tim(int i) {
    if(vs[i]) return 0;
    vs[i] = 1;
    for(int j = 1; j <= n; j++) {
	 if (a[i][j]) {
            if(y[j]==0) {
            	  x[i] = j;
            	  y[j] = i;
            	  return true;
            }
            if(tim(y[j])) {
                x[i] = j;
                y[j] = i;
            	  return true;
            }
    	 }
    }
    return false;
}
int cal() {
	int cnt = 0;
	for (int i = 1; i <= n; i++) {
		memset(vs, 0, sizeof(vs));
		if (tim(i)) cnt++;
	}
	return cnt;
}
int main() {
    ios_base::sync_with_stdio(false); cin.tie(0);
    cin >> n;
    vector < vector <int>> edge;
    for (int i = 1; i <= n; i++) {
    	for (int j = 1; j <= n; j++) {
    		int x;
    		cin >> x;
    		edge.push_back({x, i, j});
		}
	}
	sort(edge.begin(), edge.end());
	int l = 0, r = edge.size() - 1, m, ans;
	while (l <= r) {
		memset(a, 0, sizeof(a));
		memset(x, 0, sizeof(x));
		memset(y, 0, sizeof(y));
		m = (l + r) / 2;
		for (int j = 0; j <= m; j++)
			a[edge[j][1]][edge[j][2]] = 1;
		if (cal() == n) {
			ans = edge[m][0];
			r = m - 1;
		} else  l = m + 1;
	}
	cout << ans << '\n';
}

File: sach/de3397.py
//DE3397.PY - Tran Huu Nam - hsgtin.vn - 2024
import numpy as np
N = 101
a = np.zeros((N, N), dtype=int)
vs = np.zeros(N, dtype=int)
x = np.zeros(N, dtype=int)
y = np.zeros(N, dtype=int)
n = 0
def tim(i):
    if vs[i]:
        return False
    vs[i] = 1
    for j in range(1, n + 1):
        if a[i][j]:
            if y[j] == 0:
                x[i] = j
                y[j] = i
                return True
            if tim(y[j]):
                x[i] = j
                y[j] = i
                return True
    return False
def cal():
    cnt = 0
    for i in range(1, n + 1):
        vs.fill(0)
        if tim(i):
            cnt += 1
    return cnt
def main():
    import sys
    input = sys.stdin.read
    data = input().split()
    index = 0
    n = int(data[index])
    index += 1
    edge = []
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            edge.append((int(data[index]), i, j))
            index += 1
    edge.sort()
    l, r = 0, len(edge) - 1
    ans = 0
    while l <= r:
        a.fill(0)
        x.fill(0)
        y.fill(0)
        m = (l + r) // 2
        for j in range(m + 1):
            a[edge[j][1]][edge[j][2]] = 1
        if cal() == n:
            ans = edge[m][0]
            r = m - 1
        else:
            l = m + 1
    print(ans)
if __name__ == "__main__":
    main()

